### PR TITLE
ARROW-2467: [Rust] Add generated flatbuffers IPC code

### PR DESCRIPTION
**Pull request by andygrove**
Adding generated flatbuffers code for Schema, File, Message, and Tensor. It would be nice to generate on the fly as part of the build process but this doesn't seem feasible today because:

- We would have to download and compile Flatbuffers first
- The Rust implementation of Flatbuffers has a couple minor issues that required manually fixing the generated code

### CODE REVIEW

**Comment by xhochy** [comment]
> We would have to download and compile Flatbuffers first
This will be part of the next flatbuffers release, i.e. we would then be able to skip the compilation step and use a released binary, e.g. via conda?

**Comment by wesm** [comment]
> The Rust implementation of Flatbuffers has a couple minor issues that required manually fixing the generated code

Are there any bugs to report upstream?

**Comment by andygrove** [comment]
@wesm I just filed these issues:

https://github.com/google/flatbuffers/issues/5052
https://github.com/google/flatbuffers/issues/5053
https://github.com/google/flatbuffers/issues/5054

**Comment by andygrove** [comment]
@sunchao @paddyhoran @kszucs This is ready for review

**Comment by sunchao** [comment]
Not sure if it's a good idea to add generated files in the repo. Personally, I've had bad experiences with this in the past, such as having to handle different versions of the code generating software. The approach we used in parquet-rs is to create a separate crate just to store the generated files, which may or may not be feasible here..

File: rust/src/lib.rs [hunk 1]
```diff
@@ -30,6 +30,7 @@ pub mod buffer;
 pub mod builder;
 pub mod datatypes;
 pub mod error;
+pub mod format;
```
Comment (sunchao) [inline rust/src/lib.rs]: Can we rename this module to `ipc`? following the C++ impl.

**Comment by paddyhoran** [comment]
> It would be nice to generate on the fly as part of the build process but this doesn't seem feasible today

IMHO generating on the fly is the ideal situation so it depends how long it would take to get to this end state.  

> This will be part of the next flatbuffers release, i.e. we would then be able to skip the compilation step and use a released binary, e.g. via conda?

This solves the first issue in a reasonable time frame.

The other 3 issues don't seem to be big hurdles and the `flatbuffers` devs have been responsive so it's reasonable that they will be addressed in a reasonable time-frame also.

Given all this I'm in favor of merging this PR and updating to generating on the fly once all the blockers are resolved.

**Comment by andygrove** [comment]
@sunchao I started a discussion on the mail list re adding a separate crate but not responses yet. Can we consider this as a separate PR?

@paddyhoran Could you approve please given that you are in favor of merging? Appveyor failed due to something unrelated.

**Comment by sunchao** [comment]
> @sunchao I started a discussion on the mail list re adding a separate crate but not responses yet. Can we consider this as a separate PR?

I'm OK as long as this is just a temporary solution.

**Review by paddyhoran — APPROVED** [review]
LGTM

**Pull request closed without merging**